### PR TITLE
Add load scriptfile

### PR DESCRIPTION
**Pull request by higashi000**
Add function for load script file.

This is sample plugin.
https://github.com/higashi000/dps-loadfilesample
https://github.com/higashi000/dps-loadfilesample-remotefile

### CODE REVIEW

File: vim/load.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  isDirectory: boolean;
+}
+
+export async function load(
+  denops: Denops,
+  filePathInfo: URL,
+): Promise<void> {
+  const scriptPath = await setFile(denops, filePathInfo);
+
+  await execute(denops, `source ${scriptPath}`);
```
Comment (lambdalisue) [inline vim/load.ts:14]: *should*

1. Please consider what happen if `scriptPath` contains spaces, backslashes, or some special characters. Probably `scriptPath` should be escaped properly (prob. one of `fnameescape()` or `shellescape()` is good choice)
2. Please use third attribute of the `execute` function to pass `scriptPath` variable through context. It's a bit safer choice.

File: vim/load.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+async function setFile(denops: Denops, filePathInfo: URL): Promise<string> {
+  const re = /.*\//;
+  const homeDir = await Deno.env.get("HOME") as string;
```
Comment (lambdalisue) [inline vim/load.ts:19]: *should*

Both `re` and `homeDir` are not necessary to be defined in a function level. These should be defined in a module level.

Comment (lambdalisue) [reply vim/load.ts:19]: may

It's a bit better to handle a case when `HOME` is not defined if the definition moves to the module level.

File: vim/load.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+  if (filePathInfo.protocol === "file:") {
```
Comment (lambdalisue) [inline vim/load.ts:21]: :memo:

`URL.protocol` includes the final `':'`.

https://developer.mozilla.org/en-US/docs/Web/API/URL/protocol

File: vim/load.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    const filename = filePathInfo.pathname.replace(re, "");
```
Comment (lambdalisue) [inline vim/load.ts:22]: *should*

Use `path` module of std library of Deno to extract basename part from path.

File: vim/load.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    await makePluginDir(homeDir, denops.name);
+
+    await Deno.copyFile(
+      filePathInfo.pathname,
+      `${homeDir}/.cache/denops-std/${denops.name}/${denops.name}_${filename}`,
```
Comment (lambdalisue) [inline vim/load.ts:27]: *should*

Use `path` module to construct entire path string.

Comment (lambdalisue) [reply vim/load.ts:27]: **must**

We have to support Windows as well. I don't think `~/.cache` is proper directory to cache things in Windows.

File: vim/load.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    );
+
+    return `${homeDir}/.cache/denops-std/${denops.name}/${denops.name}_${filename}`;
+  } else {
+    const filename = filePathInfo.href.replace(re, "");
+    const data = (await fetch(filePathInfo.href)).arrayBuffer();
+    await makePluginDir(homeDir, denops.name);
```
Comment (lambdalisue) [inline vim/load.ts:34]: may

You can perform 1) fetching data, 2) making plugin directory in parallel. Use `Promise.all()` then.

File: vim/load.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    await Deno.stat(`${home}/.cache/denops-std`);
+  } catch (err) {
+    await Deno.mkdir(`${home}/.cache/denops-std`);
+  }
+
+  try {
+    await Deno.stat(`${home}/.cache/denops-std/${pluginName}`);
+  } catch (err) {
+    await Deno.mkdir(`${home}/.cache/denops-std/${pluginName}`);
+  }
```
Comment (lambdalisue) [inline vim/load.ts:62]: *should*

Use `recursive` option of `Deno.mkdir` to create directories at once. You don't need to check presense of directories by `Deno.stat` as well.

**Review by lambdalisue — CHANGES_REQUESTED** [review]
- **must** - Something is wrong. You must follow comments
- *should* - Something is not good. You should follow comments or explain reason
- may - Just suggestion. It's up to you to follow or not.

**Comment by lambdalisue** [comment]
I've added `pull_request` to workflow so please re-create PR to kick CI.

File: vim/load.ts (hunk 1 above)
Comment (higashi000) [reply vim/load.ts:14]: > 1
Get out this error, if `scriptPath` contains spaces, exclamation mark and question mark.

```
Unexpected error occurred in 'loadfilesample' NotFound: File not found
at processResponse (deno:core/core.js:213:11)
at Object.jsonOpAsync (deno:core/core.js230:12)
at async Object.copyFile (deno:runtime/js/30_fs.js:50:5)
at async setFile (load.ts:25:5)                                                                                           
at async load (load.ts:12:22)                                                                                             
at async Vim.load (vim.ts:62:5)                                                                                           
at async mod.ts:18:5                                                                                                      
at async Promise.all (index 1) 
```

I will fix it, use to `shellescape()`.

File: vim/load.ts (hunk 5 above)
Comment (higashi000) [reply vim/load.ts:27]: > We have to support Windows as well. I don't think `~/.cache` is proper directory to cache things in Windows.

I think `~/AppData/Local` is the way to go in Windows.

**Pull request closed without merging**